### PR TITLE
Verify board from firmware before flashing

### PR DESCRIPTION
**Pull request by haslinghuis**
Read board info from firmware and present a warning if there is a mismatch before flashing to prevent the user flashing the wrong firmware.

- The verified firmware target is selected as default upon tab load.
- Changing port will be detected. User is still allowed to change target.
- The dialog has abort and continue so it does not interfere.
- Added some checks to avoid exceptions and make it more robust.
- Added a switch in firmware flasher to disable/enable detection on tab load or port change (needs tab reload)

Edit: just found a couple of feature request:
Fixes: https://github.com/betaflight/betaflight-configurator/issues/373
Fixes: https://github.com/betaflight/betaflight-configurator/issues/168
Fixes: https://github.com/betaflight/betaflight-configurator/issues/1682

### CODE REVIEW

**Review by McGiverGim — DISMISSED** [review]
I've not tested it, but the code seems ok. We will see how it behaves under different versions/fc when merged.

**Comment by haslinghuis** [comment]
Made some further improvement. Now users can swap FC in the tab and it will verify and change the board.
Also implementing on flash button itself.

**Comment by haslinghuis** [comment]
@McGiverGim Made some further improvement. Now users can swap FC while in firmware_flasher tab and it will verify and change the board. Most time went into debugging and testing with Matek F411 and F722.

**Review by McGiverGim — APPROVED** [review]

**Comment by haslinghuis** [comment]
Have a very intermittent issue I'm not able to reproduce but will update the PR to try to catch it and cleanup when it happens.
```
serial: connection opened with ID: 3 , Baud: 115200
firmware_flasher.js:819 Requesting board information
main.html#:1 Error in event handler: TypeError: Invalid Version: null.null.0
    at new l (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:4884)
    at I (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:8498)
    at Object.L [as gte] (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:8702)
    at MspHelper.process_data (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp/MSPHelper.js:772:28)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:229:13
    at Array.forEach (<anonymous>)
    at Object.notify (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:228:24)
    at Object._dispatch_message (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:219:14)
    at Object.read (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:186:22)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/tabs/firmware_flasher.js:837:68
MSPHelper.js:1639 code: 4 - crc failed
MspHelper.process_data @ MSPHelper.js:1639
2MSPHelper.js:1657 code: 4 - crc failed. No callback
MspHelper.process_data @ MSPHelper.js:1657
MSPHelper.js:1639 code: 4 - crc failed
MspHelper.process_data @ MSPHelper.js:1639
DevTools failed to load SourceMap: Could not load content for chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/jquery-textcomplete/dist/dist/jquery.textcomplete.min.map: System error: net::ERR_FILE_NOT_FOUND
```

**Comment by haslinghuis** [comment]
I would like one improvement on this and it's saving a `diff` of current settings before flashing. @limonspb has some code which could be reused when we make it an utility. So this could be a future improvement.

File: src/js/msp/MSPHelper.js
Comment (McGiverGim) [inline src/js/msp/MSPHelper.js:737]: This seems an ugly hack for the problem.
This only happens with this two parameters? Or the `mspProtocolVersion` is bad too when the problem happens?

If both are bad, maybe simply the MSP connection is bad at this moment for some reason, if only the api version values are bad then maybe the FC is sending bad the value for some reason.

Did you verify if the `mspProtocolVersion` was bad?

Comment (McGiverGim) [reply src/js/msp/MSPHelper.js:737]: I have seen the FC has the api version defined as constants (define) so it is difficult that it fails. Something must be wrong with the MSP connection on our part, but I've never seen a problem with the reading of this in other parts of the Configurator.

Comment (haslinghuis) [reply src/js/msp/MSPHelper.js:737]: Not sure about this either as former code relies on this value. But still the ugly hack is working. There seems to be something strange with the introduction of modules.

Comment (haslinghuis) [reply src/js/msp/MSPHelper.js:737]: The error also can be found at:
```
PortHandler - Removed: [{"path":"/dev/ttyACM0","displayName":"Betaflight STM32F411"}]
DevTools failed to load SourceMap: Could not load content for chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/jquery-textcomplete/dist/dist/jquery.textcomplete.min.map: System error: net::ERR_FILE_NOT_FOUND
port_handler.js:159 PortHandler - Found: [{"path":"/dev/ttyACM1","displayName":"Betaflight STM32F411"}]
serial_backend.js:69 serial: connecting to: /dev/ttyACM1
serial.js:110 serial: connection opened with ID: 4 , Baud: 115200
serial_backend.js:249 Requesting configuration data
main.html#:1 Error in event handler: TypeError: Invalid Version: null.null.0
    at new l (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:4884)
    at I (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:8498)
    at Object.L [as gte] (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/semver-min/umd/index.js:1:8702)
    at MspHelper.process_data (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp/MSPHelper.js:772:28)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:229:13
    at Array.forEach (<anonymous>)
    at Object.notify (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:228:24)
    at Object._dispatch_message (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:219:14)
    at Object.read (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/msp.js:186:22)
    at read_serial (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/serial_backend.js:590:13)
msp.js:347 MSP data request timed-out: 4 direction: 1 tab: landing
```

Removed the hack for now:
```
    if (!FC.CONFIG.apiVersion || FC.CONFIG.apiVersion === 'null.null.0') {
        // something is wrong
        FC.CONFIG.apiVersion = '0.0.0';
    }
```
Without the hack the user has to restart configurator.

Comment (haslinghuis) [reply src/js/msp/MSPHelper.js]: Made an issue for this: https://github.com/betaflight/betaflight-configurator/issues/2498

Comment (haslinghuis) [reply src/js/msp/MSPHelper.js]: Created an issue for this https://github.com/betaflight/betaflight-configurator/issues/2498

**Comment by McGiverGim** [comment]
@haslinghuis sorry, the weekend I've been out. I've seen you have removed the "hack". This does imply that it has been fixed in some way?
If not, what is the result for the user? Not selecting the correct FC?

**Comment by haslinghuis** [comment]
The hack is still needed but is happening on connect button also. When it happens you will have to restart configurator
Added code to prevent a secondary connection. Added code to refresh the tab If the board couldn't get read somehow, but this results in a never ending loop if the user should forget to apply custom defaults. Could do one on two retries which seems sufficient.

**Comment by McGiverGim** [comment]
I suppose this happens just when connecting, we start to send MSP commands and this must be the first of them... maybe if we simply add a little pause before start sending commands it is fixed? Maybe something is not really ready at the instant zero after MSP connecting. Can we add a pause before marking the MSP as ready/the FC as connected?

**Comment by haslinghuis** [comment]
I'll let you know when it's finished. Root cause seems to be having multiple connections triggered by rapid switching between tabs or connection and flash button trying to make it fail.

**Comment by haslinghuis** [comment]
The latest changes seems to make it much harder to break, so I guess it's good to go now.

**Review by McGiverGim — APPROVED** [review]
Ok, approved. We will see if the problem persist with the user testing.

File: src/js/port_handler.js
Comment (mikeller) [inline src/js/port_handler.js:175]: Re-using this isn't very clean - better introduce an extra boolean value like `boardNeedsVerification`.

File: src/js/tabs/firmware_flasher.js
Comment (mikeller) [inline src/js/tabs/firmware_flasher.js:803]: Not sure why we are sending `MSP_API_VERSION` here?
In general, I think it would be nice if we could have the MSP call stack for detecting the board version only once, and then use it here and in the check that is in `stm32.js` - this will make it easier to maintain.

File: src/js/port_handler.js
Comment (haslinghuis) [reply src/js/port_handler.js:175]: Fixed!

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [reply src/js/tabs/firmware_flasher.js:803]: The value `FC.CONFIG.apiVersion` is needed in the `MSP_BOARD_INFO` call
Due to changing ports / having to have a connection (?) the values in `FC.CONFIG` becomes empty,  not sure if this has anything to do with the transition to modules or if this how it works.

The board is verified:

- on tab entry
- on port change
- when going to flash I present the dialog if verified board is not the selected board.

Comment (haslinghuis) [reply src/js/tabs/firmware_flasher.js:803]: The value `FC.CONFIG.apiVersion` is needed in the `MSP_BOARD_INFO` call
Due to changing ports / having to have a connection (?) the values in `FC.CONFIG` becomes empty,  not sure if this has anything to do with the transition to modules or if this how it works.

**Comment by McGiverGim** [comment]
I think it will have been easier to have an "auto detect" button next to the target selection. I think the major part of problems of this PR are part of the auto-connection :P

**Comment by haslinghuis** [comment]
All connections problems have  been solved as far I'm aware with rigorous testing. It's rock solid now.  But happy to make the check on tab entry and upon changing FC on a switch. The check is taking place on 3 occasions. The one in `stm32.js` is the most important. Guess @mikeller want's to consolidate me this in a single place. But the problem is we have to make a connection to do the `msp` calls in each case to verify board as the user could change target in between, etc.
So I guess we should refactor `stm32.js` to include the `verifyBoard` function.

Switch could be manual or automatic, leaving only flash verification mandatory.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> PASS
 - PR is not in draft -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed (6/06/2021 23:04:25) -> PASS
 - commit count less or equal to three (1 commits) -> PASS
 - 'Don't merge' label NOT found -> PASS
 - 'Needs coordination with betaflight' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> FAIL
 - assigned to an approver (None) -> FAIL
 - approver count at least three (0 approvers) -> FAIL

**Comment by Asizon** [comment]
@haslinghuis what is the purpose of the board autodetection checkbox? I think it should be enabled by default. This feature will avoid many basic user errors by installing the wrong target, better that it is always activated.

In the other hand, this feature does not work when the port is already DFU, perhaps the button that @McGiverGim  commented can be useful to exit dfu, read the board, and re-enter dfu.

**Comment by haslinghuis** [comment]
@Asizon removed the last changes (the selection of this feature).

@Asizon it's not that easy to implement but I like the idea nonetheless :)
Will look into the possibility to get out of DFU, get board info and put the board back in DFU. But this button could be problematic if board could not get out of DFU (no firmware available, or simply needs reboots or needs apply custom defaults first). Maybe we could look at a feature PR for this.

**Review by Asizon — APPROVED** [review]

**Review by McGiverGim — DISMISSED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by Asizon — APPROVED** [review]

**Review by McGiverGim — APPROVED** [review]
About the go out of dfu, is problematic because some user maybe has used the boot button to get to this state because something is wrong with the FC. If we automatically go out maybe we can't enter into DFU again.
This is one of the reasons why I thought a button is better than an "auto" detection. But at this point we can go with this.

**Review by blckmn — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - Github identifies PR as mergeable -> FAIL
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Comment by blckmn** [comment]
AUTOMERGE: (PASS)
 - Github identifies PR as mergeable -> PASS
 - Assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - 'Don't merge' label NOT found -> PASS
 - At least one 'RN:' label found -> PASS
 - 'Tested' label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Comment by mikeller** [comment]
Looks like this does not work properly: About 1 time out of five that 'Flash Firmware' is clicked (with locally loaded firmware) the flashing process does not start:

image

@haslinghuis: Can you please look into fixing this?

**Comment by mikeller** [comment]
@McGiverGim:
> About the go out of dfu, is problematic because some user maybe has used the boot button to get to this state because something is wrong with the FC. If we automatically go out maybe we can't enter into DFU again.

I totally agree - if the flight controller is already in DFU mode then we don't want to attempt auto-detection, as there is a good chance that the installed firmware isn't working, and the user has used the boot button to get to this point.

**Comment by Asizon** [comment]
@mikeller im thinking that maybe this issue is related to the new configurable port detection timer code. @haslinghuis

**Comment by haslinghuis** [comment]
@Asizon those setting are used only when set explicite in option tab. @mikeller do you have the console.log when this happens. Probably best to (re)move the lock while verifying.

**Comment by Asizon** [comment]
> @McGiverGim:
> 
> > About the go out of dfu, is problematic because some user maybe has used the boot button to get to this state because something is wrong with the FC. If we automatically go out maybe we can't enter into DFU again.
> 
> I totally agree - if the flight controller is already in DFU mode then we don't want to attempt auto-detection, as there is a good chance that the installed firmware isn't working, and the user has used the boot button to get to this point.

Ummm, most of Betaflight users continues using "bl" command to enter dfu, also if they are using correct target, they uses BL. I dont know why but they doesnt know about auto dfu feature yet :(

**Comment by haslinghuis** [comment]
@mikeller do you maybe have the console.log when this happens? Did you try to press the Flash button a second time?

Also the buttons in the screenshot should be like this -  this bug occurs due to an unknown state from the previous (flashing) operation.
Screenshot from 2021-06-29 22-24-24

**Pull request merged**
